### ISSUE TITLE
Maybe an error occurred in torch.matmul and torch.mm when pytorch == 1.9.0

### ISSUE THREAD

**Issue by YqunYang** [pytorch/pytorch#61644]
For example:
```
t = torch.zeros(15,20).cuda()
M = torch.zeros(10,5).cuda()
print(t.shape,M.shape)
t.matmul(M)
print(t.matmul(M).shape)
```
results:
```
torch.Size([15, 20]) torch.Size([10, 5])
torch.Size([15, 5])
```
But the results should be showing the mismatched error.
When we remove .cuda(), it can correctly show the mismatched error.

I had evaluated the same codes (with .cuda()) in pytorch==1.5.0, it shows mat1 and mat2 are mismatched.
----> RuntimeError: mat1 and mat2 shapes cannot be multiplied (15x20 and 10x5)

cc @ezyang @gchanan @zou3519 @bdhirsh @jbschlosser @anjali411 @ngimel

**Comment by zou3519** [comment]
This is clearly incorrect behavior so I marked it as high-priority. I am able to reproduce this on PyTorch 1.9 (but I don't have a build of PyTorch master handy to test this out).

**Comment by ngimel** [comment]
This is a duplicate of #61291, fixed on master in #61394.
